src/treemodel/selection.js

```diff
@@ -212,54 +274,230 @@ export default class Selection {
 	 */
 	setAttributesTo( attrs ) {
 		this._attrs = utils.toMap( attrs );
+		this._setStoredAttributesTo( this._attrs );
 	}
 
 	/**
-	 * Removes an attribute with given key from the selection.
+	 * Converts given range to {@link core.treeModel.LiveRange} and adds it to internal ranges array. Throws an error
+	 * if given range is intersecting with any range that is already stored in this selection.
+	 *
+	 * @private
+	 * @param {core.treeModel.Range} range Range to add.
+	 */
+	_pushRange( range ) {
+		for ( let i = 0; i < this._ranges.length ; i++ ) {
+			if ( range.isIntersecting( this._ranges[ i ] ) ) {
+				/**
+				 * Trying to add a range that intersects with another range from selection.
+				 *
+				 * @error selection-range-intersects
+				 * @param {core.treeModel.Range} addedRange Range that was added to the selection.
+				 * @param {core.treeModel.Range} intersectingRange Range from selection that intersects with `addedRange`.
+				 */
+				throw new CKEditorError(
+					'selection-range-intersects: Trying to add a range that intersects with another range from selection.',
+					{ addedRange: range, intersectingRange: this._ranges[ i ] }
+				);
+			}
+		}
+
+		this._ranges.push( LiveRange.createFromRange( range ) );
+	}
+
+	/**
+	 * Iterates through all attributes stored in current selection's parent.
 	 *
+	 * @returns {Iterable.<*>}
+	 */
+	*_getStoredAttributes() {
+		const selectionParent = this.getFirstPosition().parent;
+
+		if ( this.isCollapsed && selectionParent.getChildCount() === 0 ) {
+			for ( let attr of selectionParent.getAttributes() ) {
+				if ( attr[ 0 ].indexOf( storePrefix ) === 0 ) {
+					const realKey = attr[ 0 ].substr( storePrefix.length );
+
+					yield [ realKey, attr[ 1 ] ];
+				}
+			}
+		}
+	}
+
+	/**
+	 * Removes attribute with given key from attributes stored in current selection's parent node.
+	 *
+	 * @private
 	 * @param {String} key Key of attribute to remove.
-	 * @returns {Boolean} `true` if the attribute was set on the selection, `false` otherwise.
 	 */
-	removeAttribute( key ) {
-		return this._attrs.delete( key );
+	_removeStoredAttribute( key ) {
+		const selectionParent = this.getFirstPosition().parent;
+
+		if ( this.isCollapsed && selectionParent.getChildCount() === 0 ) {
+			const storeKey = Selection._getStoreAttributeKey( key );
+
+			this._document.enqueueChanges( () => {
+				this._document.batch().removeAttr( storeKey, selectionParent );
+			} );
+		}
 	}
 
 	/**
-	 * Removes all attributes from the selection.
+	 * Stores given attribute key and value in current selection's parent node if the selection is collapsed and
+	 * the parent node is empty.
+	 *
+	 * @private
+	 * @param {String} key Key of attribute to set.
+	 * @param {*} value Attribute value.
 	 */
-	clearAttributes() {
-		this._attrs.clear();
+	_storeAttribute( key, value ) {
```

Comment:
1) Well, we could get attribute value for selection just basing on it's key, but it is an additional call and we have that value anyway.

2) We could have just one method which would remove all stored attributes and set all current attributes. But that would mean additional reads of keys / values and additional batches / operations. I don't think it's worth it.